Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/BaseSyncTool.kt

```diff
@@ -0,0 +1,99 @@
+package de.rki.coronawarnapp.diagnosiskeys.download
+
+import de.rki.coronawarnapp.appconfig.KeyDownloadConfig
+import de.rki.coronawarnapp.diagnosiskeys.server.LocationCode
+import de.rki.coronawarnapp.diagnosiskeys.storage.CachedKey
+import de.rki.coronawarnapp.diagnosiskeys.storage.CachedKeyInfo
+import de.rki.coronawarnapp.diagnosiskeys.storage.KeyCacheRepository
+import de.rki.coronawarnapp.storage.DeviceStorage
+import timber.log.Timber
+
+open class BaseSyncTool(
+    private val keyCache: KeyCacheRepository,
+    private val deviceStorage: DeviceStorage,
+    private val tag: String
+) {
+
+    internal suspend fun invalidateCachedKeys(invalidatedKeyFiles: Collection<KeyDownloadConfig.InvalidatedKeyFile>) {
+        if (invalidatedKeyFiles.isEmpty()) {
+            Timber.tag(tag).d("No invalid files to delete.")
+            return
+        }
+
+        val badEtags = invalidatedKeyFiles.map { it.etag }
+        val toDelete = keyCache.getAllCachedKeys()
+            .filter { badEtags.contains(it.info.etag) }
+
+        Timber.tag(tag).w("Deleting invalidated cached keys: %s", toDelete.joinToString("\n"))
+        keyCache.delete(toDelete.map { it.info })
+    }
+
+    internal suspend fun requireStorageSpace(data: List<LocationData>): DeviceStorage.CheckResult {
+        val requiredBytes = data.fold(0L) { acc, item ->
+            acc + item.approximateSizeInBytes
+        }
+        Timber.tag(tag).d("%dB are required for %s", requiredBytes, data)
+        return deviceStorage.requireSpacePrivateStorage(requiredBytes).also {
+            Timber.tag(tag).d("Storage check result: %s", it)
+        }
+    }
+
+    // All cached files that are no longer on the server are considered stale
+    internal fun List<CachedKey>.findStaleData(
+        availableData: List<LocationData>
+    ): List<CachedKey> = filter { (cachedKey, _) ->
+        // Is there a day on the server that matches our cached keys day?
+        val serverHasMatchingDay = availableData
+            .mapNotNull { it as? LocationDays }
+            .any { it.dayData.contains(cachedKey.day) }
+
+        when {
+            cachedKey.type == CachedKeyInfo.Type.LOCATION_DAY -> {
+                // If there is no matching day on the server, our cached key is stale
+                return@filter !serverHasMatchingDay
+            }
+            cachedKey.type == CachedKeyInfo.Type.LOCATION_HOUR && serverHasMatchingDay -> {
+                // A cached hour for which a server day exists, means we don't need the hour anymore
+                // If there is no match, then we can't decide yet, and need to check the server for hours
+                return@filter true // Stale
```

Comment:
Hour packages should only be discarded once the Day Package has been downloaded.

Reply:
> Hour packages should only be discarded once the Day Package has been downloaded.

What's the relation to the line of code you commented on?

Reply:
_cached hour for which a server day exists, means we don't need the hour anymore_ suggests that the Hour Package is deleted _before_ the Day Package is downloaded. It should be deleted after the Day Package has been downloaded successfully.